### ISSUE TITLE
Missing memory barriers

### ISSUE THREAD

**Issue by talex5** [mirage/ocaml-vchan#122]
All the memory barriers in the vchan source have been commented out. e.g.

https://github.com/mirage/ocaml-vchan/blob/8d2b940acb9b1ff9c18424bf50e3f24bfb12a82e/lib/endpoint.ml#L245

The log message (https://github.com/mirage/ocaml-vchan/commit/0375089e2a789bee9ca5eb430199be41dc4fa8b0) from 2014 says:

> The memory barriers are in the xenctrl module at the moment which
seems the wrong place. Additionally, x86 is strongly ordered, so
comment them out for now until we can get the mb operations into a
library elsewhere.

If this library only works reliably on x86, we should probably document that somewhere.

**Comment by avsm** [comment]
Yes indeed. This strongly suggests that the library should be x86_64 only.  It would be a nicer fix to dig out the right memory barriers from upstream vchan implementations and put the right ones in this library.
